File: src/python/xerxes/extensions/skill_sources/github.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""GitHub-backed skill source.

Tests avoid hitting GitHub by injecting a ``fetch_callable`` that returns the
raw markdown body for a given ``identifier``.
"""

from __future__ import annotations

from collections.abc import Callable

from .base import SkillBundle, SkillSearchHit, SkillSource

FetchCallable = Callable[[str], str]
SearchCallable = Callable[[str, int], list[dict]]


class GitHubSkillSource(SkillSource):
    """Skill source that loads ``SKILL.md`` files from a GitHub repository."""

    name = "github"

    def __init__(
        self,
        *,
        repo: str = "agentskills/community",
        fetch_callable: FetchCallable | None = None,
        search_callable: SearchCallable | None = None,
    ) -> None:
        """Bind the source to ``repo`` with optional fetch/search callables.

        Args:
            repo: ``owner/name`` of the GitHub repository.
            fetch_callable: Function returning markdown body for an identifier.
            search_callable: Function returning raw search rows for a query.
        """
        self._repo = repo
        self._fetch_fn = fetch_callable
        self._search_fn = search_callable

    def search(self, query: str, *, limit: int = 20) -> list[SkillSearchHit]:
        """Invoke the configured search callable and adapt rows to hits."""
        if self._search_fn is None:
            return []
        raw = self._search_fn(query, limit)
        out: list[SkillSearchHit] = []
        for row in raw:
            out.append(
                SkillSearchHit(
                    name=row["name"],
                    description=row.get("description", ""),
                    source_name=self.name,
                    version=row.get("version", ""),
                    tags=list(row.get("tags", [])),
                )
            )
        return out

    def fetch(self, identifier: str) -> SkillBundle:
        """Build a ``SkillBundle`` from the configured fetch callable.

        Raises:
            RuntimeError: No fetch callable was provided at construction time.
        """
        if self._fetch_fn is None:
            raise RuntimeError("GitHubSkillSource was not configured with a fetch callable")
        body = self._fetch_fn(identifier)
        return SkillBundle(
            name=identifier,
            version="github",
            body_markdown=body,
            metadata={"repo": self._repo},
            source_name=self.name,
        )


__all__ = ["FetchCallable", "GitHubSkillSource", "SearchCallable"]

File: src/typescript/src/extensions/skillSources/github.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

import {
  normalizeSkillSearchLimit,
  requireSkillIdentifier,
  requireSkillSearchQuery,
  SkillSourceConfigurationError,
  SkillSourceError,
  type SkillBundle,
  type SkillSearchHit,
  type SkillSource,
} from './base.js'

export const DEFAULT_GITHUB_SKILL_REPOSITORY = 'agentskills/community'

export interface GitHubSkillFetchRequest {
  readonly identifier: string
  readonly repository: string
}

export interface GitHubSkillSearchRequest {
  readonly limit: number
  readonly query: string
  readonly repository: string
}

export interface GitHubSkillSearchRow {
  readonly description?: string
  readonly name: string
  readonly tags?: readonly string[]
  readonly version?: string
}

/** Host-owned GitHub integration. It may use an API client, a cache, or an approved local mirror. */
export interface GitHubSkillSourcePorts {
  readonly fetchSkillMarkdown?: (request: GitHubSkillFetchRequest) => string | Promise<string>
  readonly searchSkills?: (
    request: GitHubSkillSearchRequest,
  ) => readonly GitHubSkillSearchRow[] | Promise<readonly GitHubSkillSearchRow[]>
}

export interface GitHubSkillSourceOptions {
  readonly ports?: GitHubSkillSourcePorts
  readonly repository?: string
}

/** GitHub-backed source that performs no HTTP, authentication, or shell work itself. */
export class GitHubSkillSource implements SkillSource {
  readonly name = 'github'
  readonly repository: string
  private readonly ports: GitHubSkillSourcePorts

  constructor(options: GitHubSkillSourceOptions = {}) {
    const repository = options.repository ?? DEFAULT_GITHUB_SKILL_REPOSITORY
    if (typeof repository !== 'string' || !repository.trim()) {
      throw new SkillSourceError('GitHub skill repository must be a non-empty string')
    }
    this.repository = repository
    this.ports = options.ports ?? {}
  }

  async search(query: string, limit = 20): Promise<readonly SkillSearchHit[]> {
    const normalizedQuery = requireSkillSearchQuery(query)
    const normalizedLimit = normalizeSkillSearchLimit(limit)
    if (normalizedLimit === 0 || this.ports.searchSkills === undefined) return []
    const rows = await this.ports.searchSkills({
      query: normalizedQuery,
      limit: normalizedLimit,
      repository: this.repository,
    })
    return rows.slice(0, normalizedLimit).map(row => {
      if (typeof row.name !== 'string' || !row.name.trim()) {
        throw new SkillSourceError('GitHub search returned a row without a skill name')
      }
      return {
        name: row.name,
        description: row.description ?? '',
        sourceName: this.name,
        version: row.version ?? '',
        tags: [...(row.tags ?? [])],
      }
    })
  }

  async fetch(identifier: string): Promise<SkillBundle> {
    const id = requireSkillIdentifier(identifier)
    if (this.ports.fetchSkillMarkdown === undefined) {
      throw new SkillSourceConfigurationError('GitHubSkillSource was not configured with a fetchSkillMarkdown port')
    }
    const bodyMarkdown = await this.ports.fetchSkillMarkdown({ identifier: id, repository: this.repository })
    if (typeof bodyMarkdown !== 'string') {
      throw new SkillSourceError('GitHub fetchSkillMarkdown port must return a markdown string')
    }
    return {
      name: id,
      version: 'github',
      bodyMarkdown,
      metadata: { repository: this.repository },
      sourceName: this.name,
    }
  }
}
